### PR TITLE
Use CUDAdrv instead of CUDArt

### PR DESCRIPTION
**Pull request by MikeInnes**
This package only needs a couple of types anyway, but generally CUDAdrv is better maintained (we even discussed deprecating CUDArt in order to unify things a bit more).

### CODE REVIEW

File: src/CUBLAS.jl [hunk 1]
```diff
@@ -11,8 +11,11 @@ module CUBLAS
 
 importall Base.LinAlg.BLAS
 
-using CUDArt
-using CUDArt.rt.cudaStream_t
+using CUDAdrv: OwnedPtr, CuArray, CuVector, CuMatrix
+
+CuVecOrMat{T} = Union{CuVector{T},CuMatrix{T}}
+
+const cudaStream_t = Ptr{Void}
```
Comment (maleadt) [inline src/CUBLAS.jl]: why do you need this? there's `CuStream` with an `unsafe_convert` (ie. can be passed to ccall)

File: src/CUBLAS.jl [hunk 2]
```diff
@@ -11,8 +11,11 @@ module CUBLAS
 
 importall Base.LinAlg.BLAS
 
-using CUDArt
-using CUDArt.rt.cudaStream_t
+using CUDAdrv: OwnedPtr, CuArray, CuVector, CuMatrix
```
Comment (maleadt) [inline src/CUBLAS.jl:14]: `OwnedPtr` is an implementation detail that shouldn't be used downstream. for `cudacall` you can just use `Ptr`. why do you need it?

**Review by maleadt — DISMISSED** [review]
quick review, I can have a proper look tomorrow

File: src/CUBLAS.jl (hunk 1 above)
Comment (MikeInnes) [reply src/CUBLAS.jl]: I'm just going from the code that was there before, but can try `CuStream` instead.

File: src/CUBLAS.jl (hunk 2 above)
Comment (MikeInnes) [reply src/CUBLAS.jl:14]: There are some methods here that used to take `CudaPtr` -- but I'm not sure exactly why. Is `Ptr` the right alternative for those methods?

**Comment by maleadt** [comment]
wrt. `OwnedPtr`, this is the first design iteration so it could be improved, but conceptually I didn't want downstream packages to use it because it is just an implementation detail to deal with resource finalization. And because the design is bound to change, as is evident from this use case.

But as you say, some users will require manual memory management to build whatever they need. What about an `abstract Buffer` with `OwnedPtr` being the `struct CUDABuffer <: Buffer`, and by-protocol having an `unsafe_convert(::Ptr, Buffer` and some others?

**Comment by MikeInnes** [comment]
That could work. You could also expose `struct Buffer(ptr::T, len::Int)` or something similar, and guarantee the API of `Buffer` but not of the internal `ptr` type (other than that it can be passed to kernels).

Raw access to buffers seems essential, but I don't think we need anything more than that (e.g. pointer arithmetic, which we can do by storing an offset where needed).

File: src/CUBLAS.jl (hunk 1 above)
Comment (MikeInnes) [reply src/CUBLAS.jl]: Actually, this is just a typealias that's used by the generated C wrappers, so it seems reasonable to keep it just for internal convenience.

**Comment by MikeInnes** [comment]
I've tried removing `OwnedPtr` but I think it's going to be an uphill battle; CUBLAS uses `pointer(xs)` a lot as well as pointer arithmetic. It seems strange that `pointer` still returns an `OwnedPtr` if that's meant to be an implementation detail; and since it's not a subtype of `Ref` there's no interface we can use to handle it abstractly.

I think we should go ahead with this improvement (which is at least no worse than the original) and then figure out the right pointer design that handles these cases.

**Comment by maleadt** [comment]
> It seems strange that pointer still returns an OwnedPtr if that's meant to be an implementation detail

It ought to implement the same basic interface.

> and since it's not a subtype of Ref there's no interface we can use to handle it abstractly.

I thought subtyping `Ref` had other consequences, wrt. `ccall` and such. Not sure though.

> I think we should go ahead with this improvement (which is at least no worse than the original) and then figure out the right pointer design that handles these cases.

Was going to propose the same. We can iron out a buffer-like interface, or another design, when we have some more bandwidth to do so (hackathon). I'll try and keep `OwnedPtr` stable until then.

**Comment by MikeInnes** [comment]
Great, thanks a lot. I don't think the stability restrictions on OwnedPtr are too tight, since we're only really relying on it to exist and behave in a vaguely pointer-like way.

File: REQUIRE [hunk 3]
```diff
@@ -1,2 +1,2 @@
 julia 0.5
-CUDArt 0.3.0
+CUDAdrv 0.5.1
```
Comment (tkelman) [inline REQUIRE:2]: what in here wouldn't work with 0.5.0?

File: src/libcublas_types.jl [hunk 4]
```diff
@@ -100,4 +100,4 @@ try
 catch exception
```
Comment (tkelman) [inline src/libcublas_types.jl:100]: there's still a reference to CUDArt above so I believe this will always throw

File: src/libcublas.jl [hunk 5]
```diff
@@ -773,4 +773,4 @@ try
 catch exception
```
Comment (tkelman) [inline src/libcublas.jl:773]: there's another reference to CUDArt at the beginning of this try

**Pull request merged**